src/expr/node_traversal.cpp

```diff
@@ -0,0 +1,145 @@
+/*********************                                                        */
+/*! \file node_traversal.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Iterators for traversing nodes.
+ **/
+
+#include "node_traversal.h"
+
+namespace CVC4 {
+
+NodeDfsIterator::NodeDfsIterator(TNode n, bool postorder)
+    : d_stack{n},
+      d_visited(),
+      d_postorder(postorder),
+      d_current(TNode())
+{
+}
+
+NodeDfsIterator::NodeDfsIterator(bool postorder)
+    : d_stack(),
+      d_visited(),
+      d_postorder(postorder),
+      d_current(TNode())
+{
+}
+
+NodeDfsIterator& NodeDfsIterator::operator++()
+{
+  // If we were just constructed, advance to first visit
+  initializeIfUninitialized();
+
+  // Void our current visit, and advance to the next one
+  d_current = TNode();
+  advanceToNextVisit();
+  return *this;
+}
+
+NodeDfsIterator NodeDfsIterator::operator++(int)
+{
+  NodeDfsIterator copyOfOld(*this);
+  ++*this;
+  return copyOfOld;
+}
+
+TNode& NodeDfsIterator::operator*()
+{
+  // If we were just constructed, advance to first visit
+  initializeIfUninitialized();
+  Assert(!d_current.isNull());
+
+  return d_current;
+}
+
+bool NodeDfsIterator::operator==(const NodeDfsIterator& other) const
+{
+  // The stack and current node uniquely represent traversal state. We need not
+  // use the scheduled node set. We also ignore the order: users should not
+  // compare nodes of different order.
```

Comment:
Maybe assert that the orders are the same?

Reply:
Done in 814f73316

Reply:
Oh, I was actually thinking about a literal `Assert()` since we don't want people to compare iterators of different orders (i.e. it would indicate a bug in the code). What do you think?

Reply:
Ah! That's what you meant. Yes, that's better.

Reply:
Done in a0eaf3e11